### DISCUSSION TITLE
Blast+ in RAVEN

### DISCUSSION THREAD

**Discussion by parsanayyara** [SysBioChalmers/RAVEN#506] [Help me!] [answered]
Hello, 

I noticed that RAVEN uses BLAST+ to run getBlast() for reconstructions, I followed the BLAST+ installation as noted on the website and install it on the RAVEN folder, but everytime I tried running checkInstallation on RAVEN, it kept giving this error. Can anyone give suggestions on how to properly get Blast+ installed for RAVEN?

<img width="368" alt="image" src="https://github.com/SysBioChalmers/RAVEN/assets/77234398/5aa0d265-8bac-44af-8c87-4cd40b2364cf">

**Comment by mihai-sysbio** [comment]
Hi @parsanayyara! Could you specify what operating system and what version of Matlab you are using?

**Reply by parsanayyara** [reply]
MATLAB R2022b in 64-bit Windows

**Reply by mihai-sysbio** [reply, marked as answer]
The executables for BLAST+ (as packaged with RAVEN) haven't been updated in two years https://github.com/SysBioChalmers/RAVEN/tree/main/software/blast%2B. We are distributing these, so there is no need to install BLAST+ separately.  For testing purposes, could I ask you to fetch an older release [`v2.7.12`](https://github.com/SysBioChalmers/RAVEN/releases/tag/v2.7.12) and see how that one behaves?

**Reply by parsanayyara** [reply]
I will try and do that. On another note though, suppose I manage to get it RAVEN up and running, would it be possible to do a reconstruction only based on MetaCyc and can the draft reconstruction generated by RAVEN be exported into excel or is it already in COBRA-compatible SBML?

**Reply by mihai-sysbio** [reply]
Instead of risking to not be giving the fully accurate answers here, I'm pinging @edkerk.
In general, I wouldn't worry about the export. To my knowledge, the SBML is COBRA-compatible, and there is [Excel export](https://sysbiochalmers.github.io/RAVEN/doc/io/exportToExcelFormat.html) as well.

**Reply by edkerk** [reply]
Regarding your MetaCyc question, you can make a reconstruction based only on MetaCyc but I would not recommend this as MetaCyc-based reconstruction tend to contain many false positive hits. Instead, MetaCyc-based reconstruction is best used as complementary approach, where the reactions are checked and curated before added to another (draft) model.

Finally, RAVEN exports valid SBML L3V1 FBCv2 files, which are COBRA-compatible. I strongly discourage the use of Excel as intermediate file format. Excel can be great if you quickly want to look up some aspects of the model (and are a proficient Excel-user), but should not be used to make changes to the model.

**Comment by edkerk** [comment]
I cannot replicate the non-functioning BLASTP executable, you can perhaps run `blastPlusTests` to get more information.

**Comment by dimwas** [comment]
I can confirm the problem exists; strangely enough, it occurs on my workplace PC, but not on my laptop and home PC. On all of them I use 64-bit Windows and MATLAB R2023a. I ckecked two RAVEN releases (2.8.7 and 2.9.2), the result is the same. I tried to run `blastPlusTests`, the results are not very informative:

```
Name: 'blastPlusTests/testBlastPlus'
         ProcedureName: 'testBlastPlus'
             TestClass: [0×0 string]
            BaseFolder: 'D:\RAVEN\testing\unit_tests'
      Parameterization: [0×0 matlab.unittest.parameters.EmptyParameter]
    SharedTestFixtures: [1×0 matlab.unittest.fixtures.EmptyFixture]
                  Tags: {1×0 cell}

Tests Include:
    0 Parameterizations, 0 Shared Test Fixture Classes, 0 Tags.
```
I also tried

1. to reinstall RAVEN;
2. to use blast+ version from NCBI website;
3. to copy blast+ folder from a PC where it works.

It didn't work. Could you maybe suggest some steps to resolve the issue? @parsanayyara, were you able to find a solution?
Edit: I tried the release v.2.7.12 suggested in the answer - the problem persists.

**Reply by edkerk** [reply]
Please run the test with `runtests('blastPlusTests')`, assuming that RAVEN is in the MATLAB path (otherwise navigate to the `RAVEN\testing\unit_tests\` folder first).

**Reply by dimwas** [reply]
... quite silly of me, sorry for that. I ran the test, it claims that
```
Error using checkFileExistence
    File "C:\Users\username\AppData\Local\Temp\tpbbac38a6_e70a_4c5f_967a_468cfb7c0859\tmpDB.phr"
    cannot be found
```
I checked this manually - indeed, there is no such folder in Temp, and there is no such file anywhere in Temp. Could you maybe send me this file so that I can check whether it fixes the problem?

**Reply by edkerk** [reply]
Could you please print the whole output? That file should be generated by blast itself, so it probably fails already earlier in the script.

**Reply by dimwas** [reply]
Here is the full output:
```
>> runtests('blastPlusTests')
Running blastPlusTests

================================================================================
Error occurred in blastPlusTests/testBlastPlus and it did not run to completion.
    ---------
    Error ID:
    ---------
    ''
    --------------
    Error Details:
    --------------
    Error using checkFileExistence
    File "C:\Users\username\AppData\Local\Temp\tpbbac38a6_e70a_4c5f_967a_468cfb7c0859\tmpDB.phr"
    cannot
    be found
    
    
    Error in getMD5Hash (line 42)
    inputFile=checkFileExistence(inputFile);
    
    Error in getBlast (line 104)
        blastReport.dbHashes.phr{numel(blastReport.dbHashes.phr)+1}=getMD5Hash(fullfile(tmpDB,
        'tmpDB.phr'));
    
    Error in blastPlusTests>testBlastPlus (line 35)
    [actBlastStructure,actBlastReport]=getBlast(organismID,fastaFile,modelIDs,refFastaFiles,true,true);
================================================================================
.
Done blastPlusTests
__________

Failure Summary:

     Name                          Failed  Incomplete  Reason(s)
    =============================================================
     blastPlusTests/testBlastPlus    X         X       Errored.

ans = 

  TestResult with properties:

          Name: 'blastPlusTests/testBlastPlus'
        Passed: 0
        Failed: 1
    Incomplete: 1
      Duration: 1.2208
       Details: [1×1 struct]

Totals:
   0 Passed, 1 Failed (rerun), 1 Incomplete.
   1.2208 seconds testing time.
```

**Reply by edkerk** [reply]
Please try:
```
dbstop getBlast.m 103
runtests('blastPlusTests')
```
And when the function halts, give the output of `status` and `fastaFile{1}`.

Afterwards you can run `dbquit` to exit debug mode, and `dbclear all` to remove the debug stop.

**Reply by dimwas** [reply]
Status = 1 (double).
fastaFile = 1×1 cell array

    {'C:\Users\username\AppData\Local\Temp\tpe3362850_de05_4793_b18b_74e9d08f97fa'}

**Reply by edkerk** [reply]
We're getting somewhere :). If you run again the two lines above to stop the function at line 103, and then run the following commands and report the output of `tmp`:
```
[status, tmp]=system(['"' fullfile(ravenPath,'software','blast+',['makeblastdb' binEnd]) '" -in ' fastaFile{1} ' -out "' fullfile(tmpDB, 'tmpDB') '" -dbtype prot']);
tmp
```

**Reply by dimwas** [reply]
Sorry for a delay; here is the result:
```
tmp =

    'USAGE
       makeblastdb.exe [-h] [-help] [-in input_file] [-input_type type]
         -dbtype molecule_type [-title database_title] [-parse_seqids]
         [-hash_index] [-mask_data mask_data_files] [-mask_id mask_algo_ids]
         [-mask_desc mask_algo_descriptions] [-gi_mask]
         [-gi_mask_name gi_based_mask_names] [-out database_name]
         [-blastdb_version version] [-max_file_sz number_of_bytes]
         [-logfile File_Name] [-taxid TaxID] [-taxid_map TaxIDMapFile] [-version]
     
     DESCRIPTION
        Application to create BLAST databases, version 2.12.0+
     
     Use '-help' to print detailed descriptions of command line arguments
     ========================================================================
     
     Error: Too many positional arguments (1), the offending value: username\AppData\Local\Temp\tp75f3142c_05f5_4f8f_b891_9afe8d2e62ea
     Error:  (CArgException::eSynopsis) Too many positional arguments (1), the offending value: username\AppData\Local\Temp\tp75f3142c_05f5_4f8f_b891_9afe8d2e62ea
     '
```
What came into my mind - my username consists of my first and last name, divided by an empty space. Before that, they were both represented in error log, but here it was only the last name. Could it be that RAVEN cannot find the address correctly because of that? I am not sure now, but, IIRC, on the laptop and the home PC (where everything works fine), my user name does not consist of two words.

**Reply by edkerk** [reply]
Right, that is most likely the issue. I'll fix this in the next RAVEN release, but as a work around you can make the following changes to getBlast:

In any line that contains `system`, when referring to `-in` or `-query`, make sure that the next input will end up between `"`. Example, in line 102 `... '" -in ' fastaFile{1} ' -out "' ...` becomes `... '" -in "' fastaFile{1} '" -out "' ...`

**Reply by edkerk** [reply]
This also explains why `getDiamond` worked while `getBlast` failed, as in `getDiamond` the extra `"` were already implemented.

**Reply by dimwas** [reply]
Unfortunately, it didn't help. To be precise, there are 4 instances of `system` in `getBlast`, and after editing the corresponding lines look like this (here i use `code` to pinpoint my additions):
Line 102: [status, ~]=system(['"' fullfile(ravenPath,'software','blast+',['makeblastdb' binEnd]) '" -in `"`' fastaFile{1} '`"` -out "' fullfile(tmpDB, 'tmpDB') '" -dbtype prot']);

Line 118: [status, ~]=system(['"' fullfile(ravenPath,'software','blast+',['blastp' binEnd]) '" -query `"`' refFastaFiles{i} '`"` -out "' outFile '_' num2str(i) '" -db "' fullfile(tmpDB, 'tmpDB') '" -evalue 10e-5 -outfmt "10 qseqid sseqid evalue pident length bitscore ppos" -num_threads "' cores '"']);

Line 135: [status, ~]=system(['"' fullfile(ravenPath,'software','blast+',['makeblastdb' binEnd]) '" -in `"`' refFastaFiles{i} '`"` -out "' fullfile(tmpDB, 'tmpDB') '" -dbtype prot']);

Line 140: [status, ~]=system(['"' fullfile(ravenPath,'software','blast+',['blastp' binEnd]) '" -query `"`' fastaFile{1} '`"` -out "' outFile '_r' num2str(i) '" -db "' fullfile(tmpDB, 'tmpDB') '" -evalue 10e-5 -outfmt "10 qseqid sseqid evalue pident length bitscore ppos" -num_threads "' cores '"']);

I tried to run `checkInstallation`, checking BLAST+ still fails. I tried to run the commands you have suggested [previously](https://github.com/SysBioChalmers/RAVEN/discussions/506#discussioncomment-9881439) with the modified line 102, `tmp` now says:
```
tmp =

    'BLAST options error: Please provide a database name using -out
     '
```

**Reply by dimwas** [reply]
> This also explains why `getDiamond` worked while `getBlast` failed, as in `getDiamond` the extra `"` were already implemented.

Maybe I did not implement the changes correctly, then. I'll go check how it is made in `getDiamond`.

**Reply by edkerk** [reply]
You should add the " character only, not surrounded by `. You might have copy-pasted my suggestion from an email that GitHub sent out, which indeed shows these additional characters. When reading this discussion on the GitHub page, it is appropriately formatted.

**Reply by dimwas** [reply]
> You should add the " character only, not surrounded by `. 

Hm, I think I did exactly that. I will try to avoid the usage of `code`, because it is implemented with apostrophes on GitHub, and it makes our life harder. In line 102, the old construction was -in, surrounded by single quotation marks, and with a double quotation mark between a first single quotation mark and an -in. If I understand you correctly, you suggest the following construction: -in, surrounded by double quotation marks, and it all surrounded by single quotation marks; same for -out and -query. Is that correct? You might have thought that I have added apostrophes in the code, but I did not - I have merely used them here on GitHub in order to pinpoint the changes via `code` envronment.

> You might have copy-pasted my suggestion from an email that GitHub sent out, which indeed shows these additional characters.

No, I did it manually.

**Reply by edkerk** [reply]
Just copy-paste this text to replace the content of `getBlast.m`, taken from PR #551: https://github.com/SysBioChalmers/RAVEN/blob/4549f295a530b6e1408dce9aab1043dd78d86d11/external/getBlast.m

**Reply by dimwas** [reply]
I did it; unfortunately, still doesn't work. I ran again the line 102, the `message` is as previous:
```
message =

    'BLAST options error: Please provide a database name using -out
     '
```

**Reply by edkerk** [reply]
Ok, so it is definitely a problem with the space in the user name. See [here](https://www.ncbi.nlm.nih.gov/books/NBK569860/) and [here](https://www.biostars.org/p/10648/#32198). The suggested solutions do not work. This is a blast problem, there is not more I can fix on this.

**Reply by dimwas** [reply]
That's regrettable. Thank you very much for your time; I will check what else can be done. Probably the easiest solution would be to rename a Windows account.

**Reply by edkerk** [reply]
One last attempt, in line 64 and 65, add `('c:\tmp\')` after `tempname` (so `tempname('c:\tmp\');`), or whatever other non-space containing temporary folder you want to use.

**Reply by dimwas** [reply]
Some progress, but the problem in `New DB title`:
```
message =

    '
     
     Building a new DB, current time: 06/26/2024 16:08:41
     New DB name:   c:\tmp\tp3b366439_496f_4d3e_b56f_9d1fdf6b9724\tmpDB
     New DB title:  C:\Users\firstname lastname\AppData\Local\Temp\tpeb35645a_29c5_4b53_817e_704da2eb538e
     Sequence type: Protein
     Keep MBits: T
     Maximum file size: 1000000000B
     BLAST options error: File C:\Users\firstname does not exist
     '
```

**Reply by edkerk** [reply]
Make the same change on line 72 of `checkFileExistence.m`

**Reply by dimwas** [reply]
Aaand now it works :) Great, thank you; will this new `tmp` folder be deleted automatically?

**Reply by edkerk** [reply]
I don't think so, will look into this when properly implementing it in RAVEN.

**Reply by edkerk** [reply]
Solved in #551
